### DISCUSSION TITLE
get sorted elements from the iterator

### DISCUSSION THREAD

**Discussion by CMbeebop** [IfcOpenShell/IfcOpenShell#3894] [General]
Hi,

I am using IFCOpenshell to read ifc files and retrieve geometry elements using the 'geometry iterator' as it is explained here
https://blenderbim.org/docs-python/ifcopenshell-python/geometry_processing.html

I realized the ordering of the elements extracted from the iterator using iterator.get() is different in different computers. I wonder if there is an easy way to remedy this, maybe changing the iterator settings.

regards, Carlos

**Comment by aothms** [comment]
This surprises me somewhat. While it's never advertised, I thought the order was deterministic. On the other hand we have changed the internal storage of some maps to be hash maps instead of tree maps, which could have an affect, but I though the major type-based mapping is still sorted based on order in file.

Do you have more details on the different platforms (same os, same version, etc.) and reason why you're interested in this?

**Reply by CMbeebop** [reply]
Hi aothms,

thanks for your reply. 
I am working on a python script for reading ifc files and extracting a triangulation out of some of the elements in the iterator (the ones corresponding to building elements like walls, windows, doors,...). I am numbering the elements following the ordering given by the iterator and I would like this numbering to be consistent on different computers. I am creating a windows executable for the script, which has been used in my laptop running Windows 10, while the other computer is a Zeon PC using Windows 11.

regards, Carlos

**Reply by aothms** [reply]
And it's consistent between successive runs on the same platform?

Are the ifcopenshell versions used identical and recent?

**Reply by CMbeebop** [reply]
ifcopenshell version is the same yes v0.7.0   
I can tell this was not a problem when running in my laptop alone (ordering was consistent in multiple runs on my computer). I can make a small survey tomorrow and check if it is consistent on different computers using the same Windows version.

**Reply by CMbeebop** [reply]
Hello again,

I have more clues on the issue, it turns out that the problem is there even using the same PC in two consecutive runs of the windows executable on the same ifc file. I also tried using different PCs with different OS (Windows 8, 10 and 11) getting different ordering in the iterator elements.

I verified the ordering on vertices and faces forming the elements is consistent even though the ordering of elements retrieved from the iterator changes (i.e. shape = iterator.get(); vertices = shape.geometry.vert;  faces = shape.geometry.faces;)

Also checked that using one single thread (i.e. iterator = ifcopenshell.geom.iterator(settings, ifc_file, 1) rather than iterator = ifcopenshell.geom.iterator(settings, ifc_file, multiprocessing.cpu_count())) the ordering is fine in consecutive runs of the executable and also on different PCs even if they have different OS.

regards, Carlos

**Reply by aothms** [reply]
Thank you

> Also checked that using one single thread ... the ordering is fine in consecutive runs of the executable and also on different PCs even if they have different OS.

^ This is really essential information. 

Generally speaking that when using multi-threading you accept some sort of indeterminism in the output. I will keep this issue in the backlog for v0.8. I think what we can do is in multi-threaded mode preallocate the results in order, instead of append them to the results as soon as processing for some items is completed (which in multi-threaded applications is indeterminate).

**Reply by CMbeebop** [reply]
Hi aothms,

yes, it seems the problem is triggered by using the multi-threading in the iterator. In my python script I am writing the triangles to a file inside of the iterator loop. I guess one solution would be to store the BIM data instead and then make some kind of sorting with the elements using the geometry (like sorting by Z then X then Y using the first triangle point for example). that will introduce a time overhead in the script for the sorting, also extra memory allocation, but we should be getting the same output which is very convenient for my project.

Thanks for the help! looking forward for the next version of IFCOpenshell, you are doing a great job
regards, Carlos
